src/test/java/org/rulez/demokracia/pdengine/BeatTableBeatInformationTest.java

```diff
@@ -0,0 +1,133 @@
+package org.rulez.demokracia.pdengine;
+
+import static org.junit.Assert.*;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.rulez.demokracia.pdengine.BeatTable.Direction;
+import org.rulez.demokracia.pdengine.annotations.tested_behaviour;
+import org.rulez.demokracia.pdengine.annotations.tested_feature;
+import org.rulez.demokracia.pdengine.annotations.tested_operation;
+import org.rulez.demokracia.pdengine.exception.ReportedException;
+import org.rulez.demokracia.pdengine.testhelpers.CreatedDefaultChoice;
+
+public class BeatTableBeatInformationTest extends CreatedDefaultChoice {
+
+	@Before
+	public void setUp() throws ReportedException {
+		super.setUp();
+	}
+
+	@tested_feature("Supporting functionality")
+	@tested_operation("BeatTable")
+	@tested_behaviour("the beat information related to a and b can be obtained for forward and backward")
+	@Test
+	public void beatTable_with_not_defined_inputs() {
+		BeatTable beatTable = new BeatTable();
+
+		assertThrows(() -> beatTable.beatInformation(null, null, null))
+				.assertException(NullPointerException.class);
+	}
+	
+	@tested_feature("Supporting functionality")
+	@tested_operation("BeatTable")
+	@tested_behaviour("the beat information related to a and b can be obtained for forward and backward")
+	@Test
+	public void beatTable_forward_with_not_defined_choices() {
+		BeatTable beatTable = new BeatTable();
+		
+		assertThrows(() -> beatTable.beatInformation(null, null, Direction.DIRECTION_FORWARD)
+				).assertException(IllegalArgumentException.class)
+				 .assertMessageIs("Not existing 1st level key");
+	}
+	
+	@tested_feature("Supporting functionality")
+	@tested_operation("BeatTable")
+	@tested_behaviour("the beat information related to a and b can be obtained for forward and backward")
+	@Test
+	public void beatTable_forward_with_not_proper_matrix() {
```

Comment:
We are getting an internal error after playing with an interface incorrectly, where that interface probably should have been internal. There is a design error here. I need some time to think it through,

Reply:
Ok, I fiexed your other findings.

Reply:
@colorimeter  not exactly. See https://sites.google.com/site/unclebobconsultingllc/active-record-vs-objects I would like to stick to separation of code and data structures, which means that the data structures should be so simple that they do not lend themselves to such misbehaviour.
so far I have the following observations:
- now we do not plan to persist the BeatTable, but this could change
- a map with map values is probably too complex
- if I look at the specification, I would probably implemented it with an sql table in mind, which have one column for Candidate a, one for Candidate b, and two columns to store the members of the resulting pair.
- the above implementation would have some serious efficiency problems, like when searching an element, and not much have to do with a non-sparse matrix
- there are a number of other possible data structures to back the designed interface
- a design choice of this complexity is probably too much to give to a junior programmer. and our goal is to make designs which can be easily implemented by juniors. I know that perhaps all of you working on the project count as senior, this observation is not about that.
But I still don't know
- what would be the right choice for the data structure, and
- how to put it into the architecture model

Reply:
@magwas I get your point. Just take your time I would like to create quality code.

Reply:
OK, my two cents: we should create a mapping of Choices to integers 0..n-1 . With that help the BeatTable (and later the PathTable) became simply an int[n][n] array.

Reply:
Well, one solution would be to think about Matrix as a "primitive" type, and create it with its own internal representation and standard serialization. This way it could be a fully encapsulated old-style class (mixing structure and behaviour, but with some reasoning behind it), and we could build the tables out of it. A test of the idea would be whether you can persist it (e.g. with hibernate) without building any knowledge into it about the persistence framework.
I am not yet decided whether either this, or the Map<(Choice,Choice),Pair> idea is better.

Reply:
Asked it: https://codereview.stackexchange.com/questions/215960/proper-separation-of-behaviour-and-data-structure-when-the-data-structure-is-com